platform/wayland/cog-platform-wl.c

```diff
@@ -568,44 +565,64 @@ pointer_on_enter(void              *data,
                  wl_fixed_t         fixed_x,
                  wl_fixed_t         fixed_y)
 {
-    CogWlPlatform *platform = data;
-    CogWlDisplay  *display = platform->display;
+    CogWlSeat *seat = data;
 
-    display->seat.keyboard.serial = serial;
-    display->seat.pointer.serial = serial;
-    display->seat.pointer.surface = surface;
+    if (pointer != seat->pointer_obj) {
+        g_critical("%s: Got pointer %p, expected %p.", G_STRFUNC, pointer, seat->pointer_obj);
+        return;
+    }
+
+    seat->display->seat_default = seat;
+
+    CogWlPlatform *platform =
+        wl_surface_get_user_data(surface); // Platform is attached the context of the wl_surface. Check create_window();
```

Comment:
There is only one platform object, it's a singleton. Remember that you can always get it using `cog_platform_get_default()` so you don't need to attach it as user data to the surface. You can do the following here and in many other places in this patch:

```c
    CogWlPlatform *platform = (CogWlPlatform *) cog_platform_get_default();
```

Note that we don't need to use `COG_WL_PLATFORM()` to do a checked cast: because there is only one object, if the Wayland platform code is executed, we _know_ that `cog_platform_get_default()` always returns a `CogWlPlatform`.

I did the same kind of simplification in https://github.com/Igalia/cog/pull/565, for example.

Reply:
By the the way: this also means that there is no need to have a `CogWlPlatform*` saved inside `CogwlView` 😉

Reply:
Good point and I think I should do this in sake of simplicity having in account there is only one platform. The reason of passing the platform attached always as a user_data is in somehow with the idea of avoiding the use of global references. I think, for the case  of the Platform (since it is a singleton, as you mention) this is not needed and it is a overhead.

There is another justification for this user_data and it is what I wrote here: https://github.com/Igalia/cog/pull/623#discussion_r1377704180. Even with this, it is wrong  add changes for problems that are still not in the code yet. I will update the PR according your suggestions.